src/lib.rs

```diff
@@ -1,9 +1,361 @@
+        runtime: R,
+        config: config::RepoConfig<Store, Share>,
+    ) -> Running<Store, Share, R> {
+        // Here we set up periodic storage tasks to compact documents
+        todo!()
+    }
+
+    #[cfg(feature = "tokio")]
+    pub async fn run_tokio(
+        config: config::RepoConfig<Store, Share>,
+    ) -> Running<Store, Share, tokio::runtime::Handle> {
+        let runtime = tokio::runtime::Handle::current();
+        Self::run(runtime, config).await
+    }
+
+    /// Connect a stream and sink of `crate::Message`
+    ///
+    /// Driving this future will drive the connection until there is an error or one or the other
+    /// end drops
+    ///
+    /// The `direction` argument determines which side will send the initial `crate::Message::Join`
+    /// message. If the direction is `ConnDirection::Outgoing` then this side will send the
+    /// message, otherwise this side will wait for the other side to send the join message.
+    ///
+    /// # Arguments
+    ///
+    /// * source - An identifier for the source of the connection. Only
+    ///   currently used for logging and typically is a
+    ///   [`std::net::SocketAddr`]
+    /// * stream - The stream + sink of messages representing the connection
+    /// * direction - Which side of the connection `stream` is
+    ///
+    /// # Sharing
+    ///
+    /// Once the connection is established and we know the other ends peer ID
+    /// the repo will query all the documents it knows about using `Store::list` and
+    /// then send sync messages for every document for which
+    /// `Share::should_share` returns `ShareDecision::Share`
+    pub async fn connect<RecvErr, SendErr, Source>(
+        self,
+        source: Source,
+        stream: Store,
+        direction: ConnDirection,
+    ) -> Result<(), error::RunConnection<SendErr, RecvErr>>
+    where
+        RecvErr: std::error::Error,
+        SendErr: std::error::Error,
+        Store: Sink<Message, Error = SendErr> + Stream<Item = Result<Message, RecvErr>>,
+        Source: std::fmt::Debug + Eq,
+    {
+        todo!()
+    }
+
+    /// Connect a tokio io object
+    ///
+    /// This implements a simple length prefixed framing protocol and is intended for use with
+    /// stream oriented transports
+    #[cfg(feature = "tokio")]
+    pub async fn connect_tokio_io<Io, Source>(
+        self,
+        source: Source,
+        io: Io,
+        direction: ConnDirection,
+    ) -> Result<(), error::RunConnection<error::CodecError, error::CodecError>>
+    where
+        Io: tokio::io::AsyncRead + tokio::io::AsyncWrite,
+    {
+        todo!()
+    }
+
+    /// Attempt to load the document given by `id` from storage
+    ///
+    /// If the document is not found in storage then `Ok(None)` is returned.
+    ///
+    /// Note that this _does not_ attempt to fetch the document from the
+    /// network.
+    pub async fn load(&self, id: DocumentId) -> Result<Option<DocHandle>, Store::Error> {
+        todo!()
+    }
```

Comment:
Not sure about this one, I'd say the repo should load all documents from storage on startup, and make handles to them available via the [sync obersver](https://github.com/automerge/spanreed/blob/1727072171bf719687853e763f1e347a38160724/src/repo.rs#L244). 

I think we should have only two API for adding new documents to the repo by the local client:

1. Bootstrap from an (out-of-band) ID; this would first try to load it from storage, returning a "ready" handle if found, and fallback to returning a "non-ready" handle otherwise and start syncing it over the network. 
2. Create a brand new document, this would return a "ready" handle by default. Syncing would also start(although maybe it should start only after the first local edit?).

Reply:
I don't think that in general it's feasible to have all documents in memory. On busy servers that could be more than the memory the machine has available. But additionally, the point of `load` here is to provide a workflow where you first check whether you have the document locally, and then if not wait for sync with some other peer via `request`.

The reason this is useful is because checking whether something is in storage is a different kind of operation to syncing it from the network. E.g. if you're building a local first app and you know there's no network connection you still want to check storage, but there's no point asking the network. To me, having this kind of "check storage, then check the network" API is clearer about what is going on then having a state on the `DocHandle` which you now have to watch for changes in order to update the user about.